homeassistant/components/enasolar/config_flow.py

```diff
@@ -0,0 +1,223 @@
+        self._data: dict[str, Any] = {}
+
+    def _host_in_configuration_exists(self, host) -> bool:
+        """Return True if host exists in configuration."""
+        if host in enasolar_entries(self.hass):
+            return True
+        return False
+
+    def _get_serial_no(self):
+        return self._enasolar.get_serial_no()
+
+    def _get_max_output(self):
+        return self._enasolar.get_max_output()
+
+    def _get_dc_strings(self):
+        return self._enasolar.get_dc_strings()
+
+    def _get_capability(self):
+        return self._enasolar.get_capability()
+
+    def _get_name(self):
+        return self._data[CONF_NAME]
+
+    async def _try_connect(self, host):
+        kwargs = {}
+
+        try:
+            self._enasolar = pyenasolar.EnaSolar(host, **kwargs)
+            await self._enasolar.interogate_inverter()
+            return True
+
+        except Exception as e:
+            self._errors[CONF_HOST] = "cannot_connect"
+            _LOGGER.error("Connection to EnaSolar inverter '%s' failed (%s)", host, e)
+        return False
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return EnaSolarOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Step when user initializes a integration."""
+        self._errors = {}
+
+        if user_input is not None:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+            if self._host_in_configuration_exists(self._data[CONF_HOST]):
+                self._errors[CONF_HOST] = "already_configured"
+            self._data[CONF_NAME] = user_input[CONF_NAME]
+            if await self._try_connect(self._data[CONF_HOST]):
+                await self.async_set_unique_id(self._get_serial_no())
+                return await self.async_step_inverter()
+        else:
+            user_input = {}
+            user_input[CONF_NAME] = DEFAULT_NAME
+            user_input[CONF_HOST] = DEFAULT_HOST
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=user_input[CONF_HOST]): cv.string,
+                    vol.Optional(CONF_NAME, default=user_input[CONF_NAME]): cv.string,
+                }
+            ),
+            errors=self._errors,
+            last_step=False,
+        )
+
+    async def async_step_inverter(self, user_input=None):
+        """Give the user the opportunities to override inverter config."""
+
+        self._errors = {}
+        if user_input is not None:
+            if not (
+                (0 <= user_input[CONF_CAPABILITY] <= 7)
+                or (256 <= user_input[CONF_CAPABILITY] <= 263)
+            ):
+                self._errors[CONF_CAPABILITY] = "capabilty_invalid"
+            else:
+                self._data[CONF_CAPABILITY] = user_input[CONF_CAPABILITY]
```

Comment:
why not just say `self._data.update(user_input)` instead of setting each key individually?

Reply:
I wanted to be able to perform specific validity checks on some data before accepting it. But have used the construct suggested when the data is all OK.